controller/src/http_server.rs

```diff
@@ -0,0 +1,268 @@
+use crate::config::HttpOptions;
+use anyhow::Result;
+use async_stream::stream;
+use axum::{
+    extract::{Path, State},
+    headers::Host,
+    http::{header, Method, StatusCode},
+    response::{sse::Event, Sse},
+    routing::{get, post},
+    BoxError, Json, Router, TypedHeader,
+};
+use chrono::{DateTime, Utc};
+use futures::{Stream, TryStreamExt};
+use plane_core::{
+    messages::{
+        agent::{
+            BackendState, BackendStateMessage, DockerExecutableConfig, DockerPullPolicy,
+            ResourceLimits,
+        },
+        scheduler::{ScheduleRequest, ScheduleResponse},
+    },
+    nats::TypedNats,
+    types::{BackendId, ClusterName, ResourceLock},
+    Never,
+};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::{collections::HashMap, net::SocketAddr, sync::Arc, time::Duration};
+use tower_http::cors::{Any, CorsLayer};
+
+trait LogError<T> {
+    fn log_error(self, status_code: StatusCode) -> Result<T, StatusCode>;
+}
+
+impl<T> LogError<T> for Result<T> {
+    fn log_error(self, status_code: StatusCode) -> Result<T, StatusCode> {
+        match self {
+            Ok(v) => Ok(v),
+            Err(e) => {
+                tracing::error!(error = %e, "HTTP request failed.");
+                Err(status_code)
+            }
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct HttpSpawnRequest {
+    grace_period_seconds: Option<u64>,
+    lock: Option<String>,
+    #[serde(default = "bool::default")]
+    require_bearer_token: bool,
+    port: Option<u16>,
+    #[serde(default = "HashMap::default")]
+    env: HashMap<String, String>,
+    #[serde(default = "Vec::default")]
+    volume_mounts: Vec<Value>,
+}
+
+impl HttpSpawnRequest {
+    pub fn into_spawn_request(
+        &self,
+        cluster: ClusterName,
+        image: String,
+    ) -> Result<ScheduleRequest> {
+        let max_idle_secs = Duration::from_secs(self.grace_period_seconds.unwrap_or_else(|| 600));
```

Comment:
unwrap_or should be fine here, also would be good to put this in a const at the the top of the file.